distribution/lib/Standard/Base/0.0.0-dev/src/Enso_Cloud/Internal/Utils.enso

```diff
@@ -86,42 +60,53 @@ datalinks_api = cloud_root_uri + "connectors"
+    response = http_request method url body additional_headers error_handlers retries
     response.decode_as_json.catch Invalid_JSON error->
         Error.throw (Enso_Cloud_Error.Invalid_Response_Payload error)
 
 ## PRIVATE
    Performs a standard request to the Enso Cloud API,
    returning a raw response on success, but handling the error responses.
 
+   This method has built-in retry mechanism that may retry the request if it
+   failed due to a network error or access token expiration.
+
    Custom error handlers can be provided as a mapping from error codes
    (defined in the cloud project) to functions that take the full JSON payload
    and return a custom error.
-http_request (method : HTTP_Method) (url : URI) (body : Request_Body = Request_Body.Empty) (additional_headers : Vector = []) (error_handlers : Map Text (Any -> Any) = Map.empty) -> Response ! Enso_Cloud_Error =
+http_request (method : HTTP_Method) (url : URI) (body : Request_Body = Request_Body.Empty) (additional_headers : Vector = []) (error_handlers : Map Text (Any -> Any) = Map.empty) (retries : Integer = 3) -> Response ! Enso_Cloud_Error =
     all_headers = [authorization_header] + additional_headers
     as_connection_error err = Error.throw (Enso_Cloud_Error.Connection_Error err)
     response = HTTP.new.request (Request.new method url headers=all_headers body=body) error_on_failure_code=False
         . catch HTTP_Error as_connection_error
         . catch Request_Error as_connection_error
-    if response.code.is_success then response else
-        if response.code == HTTP_Status_Code.unauthorized then Error.throw Enso_Cloud_Error.Unauthorized else
-            payload = response.decode_as_text
-            ## If the payload can be parsed as JSON, contains a `code` field and this `code` is found in the error
-               handlers, a specialized handler will be used for the error. Otherwise, a generic Unexpected_Service_Error
-               will be reported, containing the payload for inspection.
-            json_payload = payload.parse_json.catch Invalid_JSON (_->Nothing)
-            error_code = json_payload.get "code"
-            handler = error_code.if_not_nothing <| error_handlers.get error_code
-            case handler of
-                Nothing -> Error.throw (Enso_Cloud_Error.Unexpected_Service_Error response.code payload)
-                _ : Function -> handler json_payload
+    if response.is_error && (retries > 0) then http_request method url body additional_headers error_handlers (retries - 1) else
+        if response.code.is_success then response else
+            is_unauthorized = (response.code == HTTP_Status_Code.unauthorized) || (response.code == HTTP_Status_Code.forbidden)
+            case is_unauthorized of
+                True -> if retries < 0 then Error.throw Enso_Cloud_Error.Unauthorized else
+                    Authentication.refresh_access_token
```

Comment:
Feels wrong to run the refresh 3 times? Small detail but seems like should fail if unauthorized and has done the refresh token.

Reply:
What other approach would you suggest?

I don't see any simpler ones, but I'm open to suggestions.

Currently the retry logic retries up to 3 times regardless of the kind of error.

What if the 1st retry fails due to network issue, second needs token refresh but also fails due to 500 server error and only the 3rd one succeeds? I don't see a difference between retrying for different reasons.

Note that if the 'refresh token' operation itself fails - that has no retry so it will fail regardless of the retry count. Shall I add the retry there too maybe?